contributing/component-validator-rules.md

```diff
@@ -0,0 +1,397 @@
+reference document for this tag. For all amp-tags, the documentation URLs
+will follow the above structure. Feel free to insert the URL even if it is not
+yet live.
+
+```
+  amp_layout: {
+    supported_layouts: FILL
+    supported_layouts: FIXED
+    supported_layouts: FIXED_HEIGHT
+    supported_layouts: FLEX_ITEM
+    supported_layouts: NODISPLAY
+    supported_layouts: RESPONSIVE
+  }
+}
+```
+
+This section adds validation rules for the various layout options available to
+amp tags. See
+[AMP HTML Layout System](https://github.com/ampproject/amphtml/blob/master/spec/amp-html-layout.md)
+to determine which options make sense for your tag.
+
+
+### Attribute Validation Options
+
+We saw a very simple example of an attribute validation rule above:
+
+```
+  attrs: {
+    name: "data-selected-cat"
+    value_regex_casei: "(oscar|chloe|bella)"
+  }
+```
+
+There are many other rule types for attribute validation, some of which we will
+explore here.
+
+```
+attrs: {
+  name: "data-selected-cat"
+}
+```
+By specifying no value rules, any value is allowed for this attribute.
+
+If your code expects certain values, it is best to specify them here it will
+produce helpful error messages for developers trying to debug their tag.
+
+```
+  value_regex: "(oscar|chloe|bella)"
+```
+
+Similar to `value_regex_casei`, but case-sensitive.
+
+
+```
+value: "oscar"
+```
+```
+value_casei: "oscar"
+```
+Specifies that only "oscar" is an allowed value, as case-sensitive and
+case-insensitive variants.
+
+
+```
+value_url: {
+  allow_empty: true
+  allow_relative: true
+  allowed_protocol: "https"
+  allowed_protocol: "http"
+}
+```
+This specifies that the attribute value must be a valid URL or an empty string.
+If an URL, it may be either "http" or "https" and may be relative. Note that in
+mant cases, you may only want to allow "https" as non-secure resources will
+generate mixed-mode warnings when displayed from the AMP Cache.
+
+Only one of:
+
+ - `value`
+ - `value_casei`
+ - `value_regex`
+ - `value_regex_casei`
+ - `value_url`
+
+may be specified for a single attribute.
+
+Unless specified, attributes are all optional. To specify that an attribute is
+mandatory, use the `mandatory` field:
+
+```
+attrs: {
+  name: "data-selected-cat"
+  mandatory: true
+}
+```
+
+You may also specify that exactly one of a set of attributes is present:
+
+```
+attrs: {
+  name: "data-selected-cat"
+  mandatory_oneof: "['data-selected-cat', 'img-src']"
+}
+attrs: {
+  name: "img-src"
+  mandatory_oneof: "['data-selected-cat', 'img-src']"
+  value_url: {
+    allowed_protocol: "https"
+  }
+}
+```
+
+## Final Note on Rules
+
+This document attempts to summarize some of the more commonly used rules for
+creating validator extended components. More complex rules are possible and new rule
+types can even be added as needed. If your goals are not met by the rules in
+this document, don't hesitate to contact the AMP developers and ask for
```

Comment:
perhaps link "don't hesitate to contact" to the discussion channels section of CONTRIBUTING:

https://github.com/ampproject/amphtml/blob/503f8766bbb097ac970c90b3aea21b6e8c98ea4d/CONTRIBUTING.md#discussion-channels